src/main/java/com/example/onlinebookstore/mapper/BookMapper.java

```diff
@@ -6,21 +6,38 @@
 import com.example.onlinebookstore.dto.book.CreateBookRequestDto;
 import com.example.onlinebookstore.model.Book;
 import com.example.onlinebookstore.model.Category;
+import com.example.onlinebookstore.repository.CategoryRepository;
+import java.util.Optional;
 import java.util.stream.Collectors;
 import org.mapstruct.AfterMapping;
+import org.mapstruct.Context;
 import org.mapstruct.Mapper;
 import org.mapstruct.MappingTarget;
 
 @Mapper(config = MapperConfig.class)
 public interface BookMapper {
+
     BookDto toDto(Book book);
 
-    Book toModel(CreateBookRequestDto bookDto);
+    Book toModel(CreateBookRequestDto bookDto,
+                 @Context CategoryRepository categoryRepository);
 
-    void updateBook(CreateBookRequestDto dto, @MappingTarget Book book);
+    void updateBook(@MappingTarget Book book, CreateBookRequestDto dto,
+                    @Context CategoryRepository categoryRepository);
 
     BookDtoWithoutCategoryIds toDtoWithoutCategories(Book book);
 
+    @AfterMapping
+    default void setCategories(@MappingTarget Book book,
+                               CreateBookRequestDto bookDto,
+                               @Context CategoryRepository categoryRepository) {
```

Comment:
It’s not good practice to pass dependency to the mapping method.  better make it on the service layer